File: src/ios/ee/unity_ads/UnityBannerHelper.swift
//
//  UnityBannerHelper.swift
//  Pods
//
//  Created by eps on 2/11/21.
//

import UnityAds

internal class UnityBannerHelper {
    private class func convertAdSizeToSize(_ adSize: CGSize) -> CGSize {
        return CGSize(width: Double(Utils.convertDpToPixels(Float(adSize.width))),
                      height: Double(Utils.convertDpToPixels(Float(adSize.height))))
    }

    private var _indexToSize: [Int: CGSize] = [:]

    init() {
        for index in 0...0 {
            let adSize = getAdSize(index)
            let size = UnityBannerHelper.convertAdSizeToSize(adSize)
            _indexToSize[index] = size
        }
    }

    func getAdSize(_ index: Int) -> CGSize {
        if index == 0 {
            return CGSize(width: 320, height: 50)
        }
        assert(false, "Invalid ad index")
        return CGSize(width: 320, height: 50)
    }

    private func getIndex(_ adSize: CGSize) -> Int {
        if adSize.width == 320, adSize.height == 50 {
            return 0
        }
        assert(false, "Invalid ad size")
        return 0
    }

    func getSize(index: Int) -> CGSize {
        guard let size = _indexToSize[index] else {
            assert(false, "Invalid ad index")
            return CGSize.zero
        }
        return size
    }

    func getSize(adSize: CGSize) -> CGSize {
        return getSize(index: getIndex(adSize))
    }
}

File: src/android/unity_ads/src/main/java/com/ee/internal/UnityBannerHelper.kt
package com.ee.internal

import android.graphics.Point
import androidx.annotation.AnyThread
import com.ee.Utils
import com.unity3d.services.banners.UnityBannerSize

class UnityBannerHelper {
    companion object {
        @AnyThread
        private fun convertAdSizeToSize(adSize: UnityBannerSize): Point {
            return Point(
                Utils.convertDpToPixel(adSize.width.toDouble()).toInt(),
                Utils.convertDpToPixel(adSize.height.toDouble()).toInt())
        }
    }

    private val _indexToSize: MutableMap<Int, Point> = HashMap()

    init {
        for (index in 0..0) {
            val adSize = getAdSize(index)
            val size = convertAdSizeToSize(adSize)
            _indexToSize[index] = size
        }
    }

    @AnyThread
    fun getAdSize(index: Int): UnityBannerSize {
        if (index == 0) {
            return UnityBannerSize(320, 50)
        }
        throw IllegalArgumentException("Invalid ad index")
    }

    @AnyThread
    private fun getIndex(adSize: UnityBannerSize): Int {
        if (adSize.width == 320 && adSize.height == 50) {
            return 0
        }
        throw IllegalArgumentException("Invalid ad size")
    }

    @AnyThread
    fun getSize(index: Int): Point {
        return _indexToSize[index] ?: throw IllegalArgumentException("Invalid ad index")
    }

    @AnyThread
    fun getSize(adSize: UnityBannerSize): Point {
        return getSize(getIndex(adSize))
    }
}